synapse/lib/storm.py

```diff
@@ -2391,6 +2409,192 @@ async def execStormCmd(self, runt, genr):
+            async for item in runt.execute():
+                await asyncio.sleep(0)
+
+    async def execStormCmd(self, runt, genr):
+
+        if not self.runtsafe:
+            mesg = 'The background query must be runtsafe.'
+            raise s_exc.StormRuntimeError(mesg=mesg)
+
+        async for item in genr:
+            yield item
+
+        opts = {
+            'user': runt.user.iden,
+            'view': runt.snap.view.iden,
+            'vars': dict(runt.vars),
+        }
+
+        query = await runt.getStormQuery(self.opts.query)
+
+        # make sure the subquery *could* have run with existing vars
+        query.validate(runt)
+
+        coro = self.execStormTask(query, opts)
+        runt.snap.core.schedCoro(coro)
+
+class ParallelCmd(Cmd):
+    '''
+    Execute part of a query pipeline in parallel.
+    This can be useful to minimize round-trip delay during enrichments.
+
+    Examples:
+        inet:ipv4#foo | parallel { $place = $lib.import(foobar).lookup(:latlong) [ :place=$place ] }
+
+    NOTE: Storm variables set within the parallel query pipelines do not interact.
+    '''
+    name = 'parallel'
+    readonly = True
+
+    def getArgParser(self):
+        pars = Cmd.getArgParser(self)
+
+        pars.add_argument('--size', default=8,
+            help='The number of parallel Storm pipelines to execute.')
+
+        pars.add_argument('query',
+            help='The query to execute in parallel.')
+
+        return pars
+
+    async def nextitem(self, inq):
+        while True:
+            item = await inq.get()
+            if item is None:
+                return
+
+            yield item
+
+    async def pipeline(self, runt, query, inq, outq):
+        try:
+            async with runt.getSubRuntime(query) as subr:
+                async for item in subr.execute(genr=self.nextitem(inq)):
+                    await outq.put(item)
+
+            await outq.put(None)
+
+        except asyncio.CancelledError: # pragma: no cover
+            raise
+
+        except Exception as e:
+            await outq.put(e)
+
+    async def execStormCmd(self, runt, genr):
+
+        size = await s_stormtypes.toint(self.opts.size)
+        query = await runt.getStormQuery(self.opts.query)
+
+        query.validate(runt)
+
+        async with await s_base.Base.anit() as base:
+
+            inq = asyncio.Queue(maxsize=size)
+            outq = asyncio.Queue(maxsize=size)
+
+            async def pump():
+                try:
+                    async for pumpitem in genr:
+                        await inq.put(pumpitem)
+                    [await inq.put(None) for i in range(size)]
+                except asyncio.CancelledError: # pragma: no cover
+                    raise
+                except Exception as e:
+                    await outq.put(e)
+
+            base.schedCoro(pump())
+            for i in range(size):
+                base.schedCoro(self.pipeline(runt, query, inq, outq))
```

Comment:
i think we need to await these tasks, especially pipeline throwing an exception leaves other coros in a weird state

Reply:
nope, they are explicitly awaited when leaving the with block :)